### PR TITLE
[backport] Bulk ops on TreeSet, TreeMap prefer keys from the left

### PR DESCRIPTION
**Pull request by retronym**
Backport of #9061
Fixes scala/bug#12039

### CODE REVIEW

**Comment by retronym** [comment]
Draft until #9061 is merged

File: test/junit/scala/collection/immutable/TreeMapTest.scala [hunk 1]
```diff
@@ -172,4 +172,30 @@ class TreeMapTest extends AllocationTest {
...
   }
+
+  @Test
+  def unionAndIntersectRetainLeft(): Unit = {
+    case class C(a: Int)(override val toString: String)
+    implicit val ordering: Ordering[C] = Ordering.by(_.a)
+    val c0l = C(0)("l")
+    val c0r = C(0)("r")
+    def assertIdenticalKeys(expected: Map[C, Unit], actual: Map[C, Unit]): Unit = {
+      val expected1, actual1 = new java.util.IdentityHashMap[C, Unit]()
```
Comment (mkeskells) [inline test/junit/scala/collection/immutable/TreeMapTest.scala]: Collection.newSetFromMap(…) maybe

File: test/junit/scala/collection/immutable/TreeMapTest.scala [hunk 2]
```diff
@@ -172,4 +172,30 @@ class TreeMapTest extends AllocationTest {
...
+      expected.keys.foreach(expected1.put(_, ()))
+      actual.keys.foreach(actual1.put(_, ()))
+      assertEquals(expected1, actual1)
+    }
+
+    // This holds in 2.13.x only
+    //assertIdenticalKeys(Map((c0l, ())), HashMap((c0l, ())).++(HashMap((c0r, ()))))
+
+    assertIdenticalKeys(Map((c0l, ())), TreeMap((c0l, ())).++(HashMap((c0r, ()))))
```
Comment (mkeskells) [inline test/junit/scala/collection/immutable/TreeMapTest.scala:194]: We don't test `+`, which should also be consistent.

RB.upd looks to be doing the right thing though

File: test/junit/scala/collection/immutable/TreeSetTest.scala [hunk 3]
```diff
@@ -213,4 +213,33 @@ class TreeSetTest extends AllocationTest{
...
+    implicit val ordering: Ordering[C] = Ordering.by(_.a)
+    val c0l = C(0)("l")
+    val c0r = C(0)("r")
+    def assertIdenticalElements(expected: Set[C], actual: Set[C]): Unit = {
+      val expected1, actual1 = new java.util.IdentityHashMap[C, Unit]()
+      expected.foreach(expected1.put(_, ()))
+      actual.foreach(actual1.put(_, ()))
+      assertEquals(expected1, actual1)
+    }
+    assertIdenticalElements(Set(c0l), HashSet(c0l).union(HashSet(c0r)))
```
Comment (mkeskells) [inline test/junit/scala/collection/immutable/TreeSetTest.scala:229]: We don't test +, which should also be consistent.

File: test/junit/scala/collection/immutable/TreeSetTest.scala [hunk 4]
```diff
@@ -213,4 +213,33 @@ class TreeSetTest extends AllocationTest{
       assertEquals(s"$rText $lText", rhs, lhs)
     }
   }
+
+  @Test
+  def unionAndIntersectRetainLeft(): Unit = {
```
Comment (mkeskells) [inline test/junit/scala/collection/immutable/TreeSetTest.scala:218]: looks like this is a SetTest, not a TreeSetTest

sort of 
```
for ( lhs <- anySetType;
        rhs <- andSetType;
        values <- testValues) {
   assertIdenticalElements(...)
```

File: src/library/scala/collection/immutable/RedBlackTree.scala [hunk 5]
```diff
@@ -850,23 +850,23 @@ private[collection] object RedBlackTree {
     }
 
   private[this] def _union[A, B](t1: Tree[A, B], t2: Tree[A, B])(implicit ordering: Ordering[A]): Tree[A, B] =
-    if((t1 eq null) || (t1 eq t2)) t2
-    else if(t2 eq null) t1
+    if((t2 eq null) || (t2 eq t1)) t1
```
Comment (mkeskells) [inline src/library/scala/collection/immutable/RedBlackTree.scala]: no real change here on line 853/4 863/4 and I find the original easier to read

**Review by mkeskells — APPROVED** [review]
Note to self - need to do some rework of the builders PR to bring this into line after commit

File: test/junit/scala/collection/immutable/TreeSetTest.scala (hunk 4 above)
Comment (retronym) [reply test/junit/scala/collection/immutable/TreeSetTest.scala:218]: The 2.13.x branch contains this sort of test but I wanted to keep the backport smaller and just backport concrete tests for the bug.

File: test/junit/scala/collection/immutable/TreeSetTest.scala (hunk 3 above)
Comment (retronym) [reply test/junit/scala/collection/immutable/TreeSetTest.scala:229]: Added.

File: test/junit/scala/collection/immutable/TreeMapTest.scala (hunk 1 above)
Comment (retronym) [reply test/junit/scala/collection/immutable/TreeMapTest.scala]: Good suggestion.

**Comment by retronym** [comment]
I've opened another 2.12.12 blocker ticket https://github.com/scala/bug/issues/12047 to fix the similar regression in `immutable.HashMap`.

File: src/library/scala/collection/immutable/RedBlackTree.scala (hunk 5 above)
Comment (retronym) [reply src/library/scala/collection/immutable/RedBlackTree.scala]: Agreed. I good here as a result of an automated refactoring sequence, but I've changed them around now.

**Pull request merged**